File: algorithms/code/leetcode/lc230_kth_smallest_element_in_a_bst/LC230KthSmallestElementInABST.java
package leetcode.lc230_kth_smallest_element_in_a_bst;

/**
 * https://leetcode.com/problems/kth-smallest-element-in-a-bst/
 * #medium
 */
public final class LC230KthSmallestElementInABST {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(final int x) {
            val = x;
        }
    }

    public int kthSmallest(final TreeNode root, final int k, final TreeNode counter) {
        if (root.left != null) {
            int result = kthSmallest(root.left, k, counter);
            if (counter.val == k) {
                return result;
            }
        }
        counter.val++;
        if (counter.val == k) {
            return root.val;
        }
        if (root.right != null) {
            int result = kthSmallest(root.right, k, counter);
            if (counter.val == k) {
                return result;
            }
        }
        return 0;
    }

    public int kthSmallest(final TreeNode root, final int k) {
        TreeNode counter = new TreeNode(0);
        return kthSmallest(root, k, counter);
    }
}
package leetcode.lc230_kth_smallest_element_in_a_bst;

import org.junit.jupiter.api.Test;

import leetcode.lc230_kth_smallest_element_in_a_bst.LC230KthSmallestElementInABST.TreeNode;

import static org.junit.jupiter.api.Assertions.assertEquals;

public final class LC230KthSmallestElementInABSTTests {
    @Test
    public void testLeft() throws Exception {
        TreeNode n1 = new TreeNode(1);
        TreeNode n2 = new TreeNode(2);
        TreeNode n3 = new TreeNode(3);
        TreeNode n4 = new TreeNode(4);
        n4.left = n3;
        n3.left = n2;
        n2.left = n1;
        assertEquals(2, new LC230KthSmallestElementInABST().kthSmallest(n4, 2));
    }

    @Test
    public void testRight() throws Exception {
        TreeNode n1 = new TreeNode(1);
        TreeNode n2 = new TreeNode(2);
        TreeNode n3 = new TreeNode(3);
        TreeNode n4 = new TreeNode(4);
        n1.right = n2;
        n2.right = n3;
        n3.right = n4;
        assertEquals(2, new LC230KthSmallestElementInABST().kthSmallest(n1, 2));
    }

    @Test
    public void testCoverage() throws Exception {
        TreeNode n1 = new TreeNode(1);
        TreeNode n2 = new TreeNode(2);
        TreeNode n3 = new TreeNode(3);
        TreeNode n4 = new TreeNode(4);
        TreeNode n5 = new TreeNode(5);
        TreeNode n6 = new TreeNode(6);
        n2.left = n1;
        n2.right = n3;
        n4.left = n2;
        n4.right = n5;
        n5.right = n6;
        assertEquals(5, new LC230KthSmallestElementInABST().kthSmallest(n4, 5));
    }
}

File: algorithms/code/leetcode/lc230_kth_smallest_element_in_a_bst/lc230_kth_smallest_element_in_a_bst.rs
// https://leetcode.com/problems/kth-smallest-element-in-a-bst/
// #medium

use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    pub fn new(
        val: i32,
        left: Option<Rc<RefCell<TreeNode>>>,
        right: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<Self>>> {
        Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
    }
}

pub struct Solution;

impl Solution {
    fn kth_smallest_internal(
        root: Option<Rc<RefCell<TreeNode>>>,
        k: i32,
        counter: &mut i32,
    ) -> i32 {
        if let Some(root) = root {
            let root = root.borrow();
            let result = Self::kth_smallest_internal(root.left.clone(), k, counter);
            if *counter == k {
                return result;
            }
            *counter += 1;
            if *counter == k {
                return root.val;
            }
            let result = Self::kth_smallest_internal(root.right.clone(), k, counter);
            if *counter == k {
                return result;
            }
        }
        0
    }

    pub fn kth_smallest(root: Option<Rc<RefCell<TreeNode>>>, k: i32) -> i32 {
        let mut counter = 0;
        Self::kth_smallest_internal(root, k, &mut counter)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_left() {
        let n1 = TreeNode::new(1, None, None);
        let n2 = TreeNode::new(2, n1, None);
        let n3 = TreeNode::new(3, n2, None);
        let n4 = TreeNode::new(4, n3, None);
        assert_eq!(2, Solution::kth_smallest(n4, 2));
    }

    #[test]
    fn test_right() {
        let n4 = TreeNode::new(4, None, None);
        let n3 = TreeNode::new(3, None, n4);
        let n2 = TreeNode::new(2, None, n3);
        let n1 = TreeNode::new(1, None, n2);
        assert_eq!(2, Solution::kth_smallest(n1, 2));
    }

    #[test]
    fn test_coverage() {
        let n1 = TreeNode::new(1, None, None);
        let n3 = TreeNode::new(3, None, None);
        let n2 = TreeNode::new(2, n1, n3);
        let n6 = TreeNode::new(6, None, None);
        let n5 = TreeNode::new(5, None, n6);
        let n4 = TreeNode::new(4, n2, n5);
        assert_eq!(5, Solution::kth_smallest(n4, 5));
    }
}
